cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1778,6 +1799,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+  Result<Datum> SelectKth(const RecordBatch& record_batch, const SelectKOptions& options,
+                          ExecContext* ctx) const {
+    Datum output;
+    RecordBatchSelecter selecter(ctx, record_batch, options, &output);
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+  Result<Datum> SelectKth(const Table& table, const SelectKOptions& options,
+                          ExecContext* ctx) const {
+    Datum output;
+    TableSelecter selecter(ctx, table, options, &output);
+    ARROW_RETURN_NOT_OK(selecter.Run());
+    return output;
+  }
+};
+
+class SelectKMetaFunction : public MetaFunction {
+ public:
+  SelectKMetaFunction()
+      : MetaFunction("select_k", Arity::Unary(), &select_k_doc, &kDefaultSelectKOptions) {
+  }
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
+    return impl.ExecuteImpl(args, options, ctx);
+  }
+};
+
+class TopKMetaFunction : public MetaFunction {
+ public:
+  TopKMetaFunction()
+      : MetaFunction("top_k", Arity::Unary(), &top_k_doc, &kDefaultTopKOptions) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
+    const TopKOptions& opts = static_cast<const TopKOptions&>(*options);
+
+    std::vector<SortKey> sort_keys;
+    for (const auto& name : opts.keys)
+      sort_keys.emplace_back(SortKey(name, TopKOptions::order()));
+    if (args[0].kind() == Datum::ARRAY || args[0].kind() == Datum::CHUNKED_ARRAY) {
+      sort_keys.emplace_back(SortKey("not-used", TopKOptions::order()));
+    }
+    SelectKOptions select_k_options(opts.k, sort_keys, opts.kind);
+    return impl.ExecuteImpl(args, &select_k_options, ctx);
+  }
+};
+
+class BottomKMetaFunction : public MetaFunction {
+ public:
+  BottomKMetaFunction()
+      : MetaFunction("bottom_k", Arity::Unary(), &bottom_k_doc, &kDefaultBottomKOptions) {
+  }
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    SelectKthMetaFunction impl;
+    const BottomKOptions& opts = static_cast<const BottomKOptions&>(*options);
+
+    std::vector<SortKey> sort_keys;
+    for (const auto& name : opts.keys)
+      sort_keys.emplace_back(SortKey(name, BottomKOptions::order()));
+    if (args[0].kind() == Datum::ARRAY || args[0].kind() == Datum::CHUNKED_ARRAY) {
+      sort_keys.emplace_back(SortKey("not-used", BottomKOptions::order()));
+    }
+    SelectKOptions select_k_options(opts.k, sort_keys, opts.kind);
+    return impl.ExecuteImpl(args, &select_k_options, ctx);
+  }
+};
```

Comment:
As stated above I would rather we not have explicit top_k and bottom_k functions if they're just going to dispatch to select_k. We should have only select_k, and top/bottom_k can be convenience constructors on the options themselves.